ReactNativeClient/lib/reducer.js

```diff
@@ -284,22 +311,38 @@ function recordLastSelectedNoteIds(state, noteIds) {
 
 function changeSelectedNotes(state, action, options = null) {
 	if (!options) options = {};
-	if (!('clearNoteHistory' in options)) options.clearNoteHistory = true;
 
 	let noteIds = [];
 	if (action.id) noteIds = [action.id];
 	if (action.ids) noteIds = action.ids;
 	if (action.noteId) noteIds = [action.noteId];
 
-	// const noteIds = 'id' in action ? (action.id ? [action.id] : []) : action.ids;
-
 	let newState = Object.assign({}, state);
 
 	if (action.type === 'NOTE_SELECT') {
-		if (JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)) return state;
 		newState.selectedNoteIds = noteIds;
 		newState.newNote = null;
 		newState.selectedNoteHash = action.hash ? action.hash : '';
+
+		let backwardHistoryNotes = newState.backwardHistoryNotes.slice();
+		let forwardHistoryNotes = newState.forwardHistoryNotes.slice();
+
+		if (action.historyAction == 'goto' && action.id != action.lastSeenNote.id) {
+			forwardHistoryNotes = [];
+			backwardHistoryNotes.push(Object.assign({}, action.lastSeenNote));
+		} else if (action.historyAction === 'pop') {
+			forwardHistoryNotes.push(Object.assign({}, action.lastSeenNote));
+			backwardHistoryNotes.pop();
+		} else if (action.historyAction === 'push') {
+			backwardHistoryNotes.push(Object.assign({}, action.lastSeenNote));
+			forwardHistoryNotes.pop();
+		}
+
+		newState.backwardHistoryNotes = backwardHistoryNotes;
+		newState.forwardHistoryNotes = forwardHistoryNotes;
+
+		if (JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)) return newState;
+
```

Comment:
It should just be `return newState`

`JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)` will always be true since we are assigning `newState.selectedNoteIds = noteIds` before the if condition.

I removed `if (JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)) return state;` because of this situation:

Suppose we are in note1. We then search for note1 again. Now when we press the back button, 
`JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)` will be true and the old state will be returned without doing the pop operation. 

The history operations should be performed irrespective of `JSON.stringify(newState.selectedNoteIds) === JSON.stringify(noteIds)`